### PR TITLE
HTTP span name: Recommend handler over route name

### PR DESCRIPTION
**Pull request by Oberon00**
Recommend handler name over route name, improve formatting (split into paragraphs).

Depends-on: #557

### CODE REVIEW

File: specification/data-http.md [hunk 1]
```diff
@@ -22,16 +22,29 @@ and various HTTP versions like 1.1, 2 and SPDY.
...
 is a user id, which creates high cardinality value space not suitable for span
-names. In case of HTTP servers, these endpoints are often mapped by the server
-frameworks to more concise _HTTP routes_, e.g. `/api/users/{user_id}`, which are
-recommended as the low cardinality span names. However, the same approach usually
-does not work for HTTP client spans, especially when instrumentation is provided
+names. In case of HTTP servers, these paths are often mapped by the server
+frameworks to more concise _HTTP routes_, e.g. `/api/users/{user_id}`.
+In turn, these routes are handled by a particular handler (function/class/etc.; e.g. `display_user`).
+
+> Note that some web frameworks call the handlers endpoints,
```
Comment (carlosalberto) [inline specification/data-http.md:32]: Typo? or is this a quote?

**Comment by carlosalberto** [comment]
Overall LGTM

File: specification/data-http.md (hunk 1 above)
Comment (arminru) [reply specification/data-http.md:32]: It's just used for formatting, it makes the note grey and indented: https://github.com/Oberon00/opentelemetry-specification/blob/patch-7/specification/data-http.md#name

File: specification/data-http.md [hunk 2]
```diff
@@ -22,16 +22,29 @@ and various HTTP versions like 1.1, 2 and SPDY.
...
+while others refer to the complete URI as the "endpoint".
+Thus, the term "endpoint" should be avoided.
+
+Because multiple routes can map to the same handler, but not vice versa,
+the handler name is recommended as the low cardinality span name.
```
Comment (yurishkuro) [inline specification/data-http.md:37]: I disagree with this. When an Ops person is looking at a trace, seeing the route (service API related concept) is likely going to be more informative than the handler method name (an implementation detail).

File: specification/data-http.md (hunk 1 above)
Comment (Oberon00) [reply specification/data-http.md:32]: Yes, it was meant as an "aside". Do you think I should format it differently?

File: specification/data-http.md (hunk 2 above)
Comment (Oberon00) [reply specification/data-http.md:37]: On the other hand, if sampling is done based on the span name, then this new name is an improvement. I think that is more a problem of the span name concept which seems to play a double-role as a human-readable name *and* a grouping key.

Still, a reasonable argument.

Comment (Oberon00) [reply specification/data-http.md:37]: However, the route is also in a semantic attribute `http.route`, and it should be possible for a backend to replace the name with that (in the UI). This could even be configured differently for ops teams and dev teams.

Comment (bogdandrutu) [reply specification/data-http.md]: Same argument can be made against this to have the handler name as convention.

**Comment by Oberon00** [comment]
Closing this, as @yurishkuro made a [good argument](https://github.com/open-telemetry/opentelemetry-specification/pull/548#discussion_r404397569) as to why the current suggestion of the route name is at least not worse than the handler name and there seems to be no traction in another direction.

File: specification/data-http.md (hunk 2 above)
Comment (Oberon00) [reply specification/data-http.md]: I closed this PR and opened https://github.com/open-telemetry/opentelemetry-specification/issues/557 to track the span name double bind issue.

**Comment by Oberon00** [comment]
The current discussion in #557 seems to lean towards low cardinality being more important and we should provide a new semantic convention for the display name if we need one. Thus I'm reopening this for discussion.

**Comment by bogdandrutu** [comment]
@Oberon00 should this PR be removed?

**Comment by Oberon00** [comment]
If that's what you mean, the re-open was intentional, see the comment above. I would rather merge this PR than remove it 😃

**Comment by Oberon00** [comment]
@yurishkuro With #557 we would use the span name only as a default grouping. The route is still available as `http.route`. And we could add a `display_name` thing later. As #557 says, using the span name both as a display name and a grouping key is bad, so this PR ignores the usability of the span name as a display name and improves its value as a grouping key.

This is what the spec currently already says about the span name:

> The span name should be **the most general** string that identifies a (statistically) interesting class of Spans, rather than individual Span instances.

(emphasis added by me).

File: specification/data-http.md (hunk 2 above)
Comment (Oberon00) [reply specification/data-http.md]: @bogdandrutu  The handler name will always have a lower or the same cardinality as the route, so it is strictly better as grouping key if it's known.

Comment (Oberon00) [reply specification/data-http.md]: @yurishkuro In https://github.com/open-telemetry/opentelemetry-specification/pull/730#discussion_r459694652 you argue against the display.name convention by saying that backends will compute a better display name anyway based on span attributes. Here you argue against using a particular span name because it is bad as display name. It seems to me like these comments contradict each other. Did you change your opinion or am I misunderstanding you?

Comment (yurishkuro) [reply specification/data-http.md]: I didn't say the backends necessarily _would_ do that, but they _might_ do it if they wanted to. My argument against `display.name` was that in most cases it is redundant, especially in the examples provided in #730. But there may be cases when it is useful. However, it has another big issue when it comes to displaying aggregations, which I just posted in https://github.com/open-telemetry/opentelemetry-specification/pull/730#issuecomment-663622817.

I think my position remains the same - span name is both a grouping key and the display name. The additional `display.name` attribute is possible, but not well thought-through yet, it raises more questions for me now than it answers.

**Comment by bogdandrutu** [comment]
Assigned @yurishkuro to make a decision on this PR, move forward or close.

**Comment by Oberon00** [comment]
This PR needs a decision on #557 first.
Only if we decide the span name is (mostly) a "grouping key" as opposed to a display name (which is the direction the discussion on #557 has been going lately), would it make sense to merge this PR.

**Comment by iNikem** [comment]
Don't we blur the line between data collection and backend too much?

**Comment by Oberon00** [comment]
How so?

**Comment by iNikem** [comment]
Seems like we try to dictate what backend should do. Yes, we define semantics of the data we collect, but actually saying "this should be used for displaying and this for grouping" is going much farther, no?

**Comment by Oberon00** [comment]
I see it that way: If we want backends to be able to use this as grouping key, we have to generate something that is suitable as grouping key. For example a full URL would be fine as display name but usually not so much as grouping key. Of course, we won't stop backends from using the grouping key also or only as display name (for which it is usually still usable, though maybe not optimal).

**Comment by Oberon00** [comment]
@yurishkuro 
> The discussion in #557 about grouping key vs. display name does not affect my argument

Is the handler name (if available!) a more general name than the route? Yes, because you can map more than one route to the same handler. Is the handler name human readable? Yes (I agree that most will find the route more useful to read but as per #810, the generality is more important.

The purpose of #810 was precisely to have a guideline, criteria with which to avoid such a discussion.

> the HTTP route is ultimately the API exposed by the server, it's the notion that users use to think about the service, express metrics and SLAs, reason about inter-service dependencies, etc. Those are not purely display concerns. The handler name is an implementation detail that is almost irrelevant to all those concerns.

Agree on all points here, but we already have the http.route attribute for that. The span name has only one purpose: Group the largest meaningful set of spans together. If you need to track an SLA on a route, you can surely customize your sampler and backend to sample/group per http.route instead of the name. But usually if multiple routes map to the same handler, it does make sense to group them together, at least initially.

**Comment by yurishkuro** [comment]
> But usually if multiple routes map to the same handler, it does make sense to group them together, at least initially.

It may be sensible in some scenarios, but it is not unequivocal. I don't think the spec should be insisting on it. Some HTTP frameworks may not even have the notion of a handler name, it's fairly common to use anonymous lambda functions as handlers.

It is clear that we do not have consensus on this topic, so I suggest rewording the PR that using a handler name is an acceptable alternative, but stop short of saying that it is _preferred_ over HTTP route.

**Comment by Oberon00** [comment]
We already have "MAY provide hooks to allow custom logic to override the default span name" here. We should specify a clear order of preference for the default.
We also already say "HTTP spans MUST follow the overall guidelines for span names.", so even if I don't spell the preference out here, it would follow from the general guidelines saying "The span name SHOULD be the most general string that identifies a (statistically) interesting *class of Spans*".

**Comment by Oberon00** [comment]
Note that I still think this is a good idea and I will re-open this PR after GA.
